### ISSUE TITLE
Discrepancy in query between hyperlog based osm-p2p-db and kappa-osm

### ISSUE THREAD

**Issue by okdistribute** [digidem/mapeo-migrate#1]
**NOTE**: This requires the latest mapeo core with https://github.com/digidem/mapeo-core/pull/18 and  https://github.com/digidem/mapeo-core/pull/17 -- until they're merged, I cloned mapeo-core and pulled these two branches into the mainline, and used `npm link` 

```
node index.js /path/to/datafile.mapeosinangoe output
```
Gives you `old` which is the old database in raw format (unzipped/tared), and `output`, which is a kappa-osm and media folder.

Run the tests to see if they're the same:
```
node test.js old/ output/
```

Should see:
```
total versions in mapping.json 53864
total hyperlog entries 53864
total kappa entries 53864
old query: 53214
new query: 53255
```

The new query has more elements, why? T_T

**Comment by ghost** [comment]
The first thing I would check is to see if any of those documents in the bigger result are duplicates of each other.

**Comment by okdistribute** [comment]
Thanks @substack, but I wasn't able to find any duplicates in the new query.

**Comment by okdistribute** [comment]
I *did* find duplicate ids though, actually. Just not duplicate versions! Only 4 duplicates. I have the file on me.

**Comment by hackergrrl** [comment]
I looked into id `8019261586198480702` for a bit:

- it has two versions: seq 288 and seq 53043. The latter is a head, that links to the former.
- re bkd indexing operations, I see, in order: `INSERT 288`, `INSERT 53043`, `DELETE 288`
- however, the bkd query returns both, not just the head

I created a new mapeo database with JUST these two elements, and it doesn't reproduce the issue.

**Comment by hackergrrl** [comment]
Update:

- I modified the test script to print a diff of what's in old-db-only and what's in new-db-only
- Deletions weren't being handled correctly. Now there's nothing that's new-db-only, and a bunch of things in old-db-only
- The only things that are only showing up from the old db are:
  - deleted ways & nodes
  - nodes that are at `lat=0`, `lon=0`
- The duplicate ways are just regular forked heads -- so these are perfectly :ok_hand:
- I still need to look into the rest of the duplicate nodes, but they may also be related to the bkd issue above ^^^

**Comment by hackergrrl** [comment]
@substack Yeah, I *think* it's a bkd issue? If you clone this repo and follow the readme instructions, then run this script:

```js
// query.js
var path = require('path')
var Mapeo = require('@mapeo/core')
var OsmKappa = require('./kappa.js')

var mapeo = new Mapeo(OsmKappa(path.join(process.argv[2], 'data')))

var lon = -77.5101814093397
var lat = 0.23234501780439767
var size = 0.00001
var bbox = [lon - size, lat - size, lon + size, lat + size]

mapeo.osm.query(bbox, console.log)
```

as `node query.js output/`, you'll see that two of the nodes returned are the same ID, but only one is a head.

Oh! You'll also need the dataset. I think @karissa has a google drive link she can maybe share?
